Incorrect boundary events behavior when hit test target changes under stationary pointing device

lilyspiniolas opened the pull request:
#### edb7dbd59d799e010a5acdcea6a0821df9f14a79
<pre>
Incorrect boundary events behavior when hit test target changes under stationary pointing device
<a href="https://bugs.webkit.org/show_bug.cgi?id=298763">https://bugs.webkit.org/show_bug.cgi?id=298763</a>
<a href="https://rdar.apple.com/160147423">rdar://160147423</a>

Reviewed by Abrar Rahman Protyasha.

We usually fail to send boundary events when the hit test target underneath a stationary pointing
device changes due to layout updates. The only exception is when the pointer first moves into an
element, and that element is then hidden. If the element is then shown and hidden again while
the pointer remains stationary, boundary events will only fire on the first hide. Boundary
events do eventually fire, but only when the mouse moves.

During post layout tasks, we now schedule mouse event target updates which checks if the mouse
is still over the same element it was previously. If the element has changed, we call
`updateMouseEventTargetNode` with a synthetic `PlatformMouseEvent` so that the boundary events are
dispatched, and we update the hover state of the element the mouse is over as well.

As a result of these changes, we no longer need to call `dispatchFakeMouseMoveEventSoon` during
style changes.

The web-platform-test `pointerevent_pointerout_no_pointer_movement.html` now passes when run using
webdriver as a result, but the test still times out in WKTR ever since
<a href="https://github.com/web-platform-tests/wpt/commit/01186a0a3449e8c7f89553abe2524eced3bc30a5">https://github.com/web-platform-tests/wpt/commit/01186a0a3449e8c7f89553abe2524eced3bc30a5</a> for
reasons not yet determined. A new layout test has been added to test this behavior in the meantime:
pointerevents/mouse/pointer-boundary-events.html

Test: pointerevents/mouse/pointer-boundary-events.html
* LayoutTests/fast/events/ghostly-mousemoves-in-subframe-expected.txt: Renamed from LayoutTests/platform/glib/fast/events/ghostly-mousemoves-in-subframe-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_boundary_events_attributes_during_drag-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointerout_no_pointer_movement-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/layout_change_should_fire_mouseover-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_reappending_last_over_target.tentative-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/fast/events/ghostly-mousemoves-in-subframe-expected.txt: Removed.
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/win/fast/events/ghostly-mousemoves-in-subframe-expected.txt: Removed.
* LayoutTests/pointerevents/mouse/pointer-boundary-events-expected.txt: Added.
* LayoutTests/pointerevents/mouse/pointer-boundary-events.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::EventHandler):
(WebCore::EventHandler::~EventHandler):
(WebCore::EventHandler::clear):
(WebCore::getElementOrAncestorElementForNode):
(WebCore::EventHandler::updateMouseEventTargetNode):
(WebCore::EventHandler::scheduleMouseEventTargetUpdateAfterLayout):
(WebCore::EventHandler::updateMouseEventTargetAfterLayoutIfNeeded):
(WebCore::EventHandler::hoverTimerFired):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::performPostLayoutTasks):

Canonical link: <a href="https://commits.webkit.org/300277@main">https://commits.webkit.org/300277@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd17e82814c3d612d09eb263628bacb21ddfd13a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121989 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41691 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32361 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128548 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74082 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f681972d-d7a4-42f1-b91e-a77cb6c736e3) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123865 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42406 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50285 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92709 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61597 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/52107220-0684-45eb-b677-07121587077a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124941 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33811 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109242 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73365 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/daa6790c-d01b-40ff-a246-6b27851cc519) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32823 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27404 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72046 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103319 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/27599 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131313 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48928 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/37202 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101268 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49302 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105457 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101139 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25646 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46510 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24626 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45629 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48785 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54519 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48255 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51605 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49935 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -3052,6 +3066,68 @@ void EventHandler::clearElementUnderMouse()
+    m_mouseEventTargetFinalUpdateTimer.startOneShot(200_ms);
+}
+
+static constexpr OptionSet<HitTestRequest::Type> hoverTimerHitType { HitTestRequest::Type::Move, HitTestRequest::Type::DisallowUserAgentShadowContent };
```
CTAD
```suggestion
static constexpr OptionSet hoverTimerHitType { HitTestRequest::Type::Move, HitTestRequest::Type::DisallowUserAgentShadowContent };
```

aprotyas commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2897,13 +2897,6 @@ void Document::resolveStyle(ResolveStyleType type)
         if (m_renderView->needsLayout())
             frameView->layoutContext().scheduleLayout();
 
-        // As a result of the style recalculation, the currently hovered element might have been
-        // detached (for example, by setting display:none in the :hover style), schedule another mouseMove event
-        // to check if any other elements ended up under the mouse pointer due to re-layout.
-        RefPtr localMainFrame = this->localMainFrame();
-        if (m_hoveredElement && !m_hoveredElement->renderer() && localMainFrame)
-            localMainFrame->eventHandler().dispatchFakeMouseMoveEventSoon();
```
Is this covered by the timer callbacks?

aprotyas replied:
and/or I think I did not understand the motivation here

lilyspiniolas replied:
Yes, it is

lilyspiniolas committed:
Incorrect boundary events behavior when hit test target changes under stationary pointing device
https://bugs.webkit.org/show_bug.cgi?id=298763
rdar://160147423

Reviewed by Abrar Rahman Protyasha.

We usually fail to send boundary events when the hit test target underneath a stationary pointing
device changes due to layout updates. The only exception is when the pointer first moves into an
element, and that element is then hidden. If the element is then shown and hidden again while
the pointer remains stationary, boundary events will only fire on the first hide. Boundary
events do eventually fire, but only when the mouse moves.

During post layout tasks, we now schedule mouse event target updates which checks if the mouse
is still over the same element it was previously. If the element has changed, we call
`updateMouseEventTargetNode` with a synthetic `PlatformMouseEvent` so that the boundary events are
dispatched, and we update the hover state of the element the mouse is over as well.

As a result of these changes, we no longer need to call `dispatchFakeMouseMoveEventSoon` during
style changes.

The web-platform-test `pointerevent_pointerout_no_pointer_movement.html` now passes when run using
webdriver as a result, but the test still times out in WKTR ever since
https://github.com/web-platform-tests/wpt/commit/01186a0a3449e8c7f89553abe2524eced3bc30a5 for
reasons not yet determined. A new layout test has been added to test this behavior in the meantime:
pointerevents/mouse/pointer-boundary-events.html

Test: pointerevents/mouse/pointer-boundary-events.html
* LayoutTests/fast/events/ghostly-mousemoves-in-subframe-expected.txt: Renamed from LayoutTests/platform/glib/fast/events/ghostly-mousemoves-in-subframe-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_boundary_events_attributes_during_drag-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointerout_no_pointer_movement-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/layout_change_should_fire_mouseover-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_reappending_last_over_target.tentative-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/fast/events/ghostly-mousemoves-in-subframe-expected.txt: Removed.
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/win/fast/events/ghostly-mousemoves-in-subframe-expected.txt: Removed.
* LayoutTests/pointerevents/mouse/pointer-boundary-events-expected.txt: Added.
* LayoutTests/pointerevents/mouse/pointer-boundary-events.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::EventHandler):
(WebCore::EventHandler::~EventHandler):
(WebCore::EventHandler::clear):
(WebCore::getElementOrAncestorElementForNode):
(WebCore::EventHandler::updateMouseEventTargetNode):
(WebCore::EventHandler::scheduleMouseEventTargetUpdateAfterLayout):
(WebCore::EventHandler::updateMouseEventTargetAfterLayoutIfNeeded):
(WebCore::EventHandler::hoverTimerFired):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::performPostLayoutTasks):

Canonical link: https://commits.webkit.org/300277@main
